Add grafeas in-toto interface

kristelfung opened the pull request:
Add totoify-grafeas library w/ unit tests!

adityasaky committed:
Add interface to convert between Grafeas and native in-toto link formats

Signed-off-by: Aditya Sirish <aditya@saky.in>

kristelfung committed:
Add tests:
- load occurence and note json files
- Create GrafeasInTotoOccurrence instance
- Update test_totoifylib.py

Signed-off-by: Kristel Fung <kristelfung@gmail.com>

adityasaky committed:
Add functionality for creating in-toto links from occurrences

Signed-off-by: Aditya Sirish <aditya@saky.in>
Signed-off-by: Kristel Fung <kristelfung@gmail.com>

kristelfung committed:
- Rename test, link -> occurrence
- Fix constructor
- Add grafeas style artifact objects
- Handle return values
- Misc fixes

If we convert all byproducts and environment fields to str, we have no path back.
It's likely better to let an in-toto to Grafeas conversion fail because of
disallowed types, as then the user will know what's supported. A blanket conversion
to str() could mean a user proceeds with the conversion with no way to revert and
verify metadata.

Signed-off-by: Aditya Sirish <aditya@saky.in>
Signed-off-by: Kristel Fung <kristelfung@gmail.com>

kristelfung committed:
- Add in-toto-verify test and related setup
- Update test metadata files and keys
- Clean up commented code

Signed-off-by: Kristel Fung <kristelfung@gmail.com>
Signed-off-by: Aditya Sirish <aditya@saky.in>

adityasaky commented:
Thanks for opening this PR, and great work so far @kristelfung! I think it'll be useful to get feedback, but I'm just going to note some things we should work on as we move forward.

- Use the [new SSL python style guide](https://github.com/secure-systems-lab/code-style-guidelines/pull/21) (this includes writing some documentation for the primary methods as well as noting special cases / edge cases such as `return-value`).
- Develop the transport - all the focus has been on the other interfaces so far.
- Add in the full demo supply chain with more links rather than just the clone step - this shouldn't be too difficult as the keys are in there already.

cc @SantiagoTorres, @lukpueh

lukpueh commented:
> Use the new SSL python style guide (this includes writing some documentation for the primary methods as well as noting special cases / edge cases such as return-value).

💯  !! (You don't need to read up on the entire discussion of that PR, it's okay to just take a look at the [new guideline document](https://github.com/secure-systems-lab/code-style-guidelines/blob/49b4aa89d72ad014d909bbef055ab534f58ec2db/python.md).)

kristelfung committed:
Add docstrings

Signed-off-by: Kristel Fung <kristelfung@gmail.com>

kristelfung committed:
- RSA key import fix
- Update test metadata
- Fix JSON output
- Update in-toto version

Signed-off-by: Kristel Fung <kristelfung@gmail.com>
Signed-off-by: Aditya Sirish <aditya@saky.in>

SantiagoTorres commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,277 @@
+
+
+
+def create_grafeas_occurrence_from_in_toto_link(in_toto_link, step_name, resource_uri):
```
I'm assuming this is a static method. I wonder if we'd like something a little more succinct for this method name. I'm thinking of:

```
grafeas_link = GrafeasInTotoOcurrence.from_link(link)
```

Should make things readable and short...

SantiagoTorres commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,277 @@
+
+
+
+def create_in_toto_link_from_grafeas_occurrence(grafeas_occurrence, step_name):
```
I think the same could be done here (but instead we have a instance method):

```
grafeasOccurrence = GrafeasInTotoOcurrence(...)
grafeasOccurrence.to_link()
```
I also think it'd be a little simpler to follow when doing something like this:

```
my_link = ...
grafeas_occurrence = GrafeasInTotoOcurrence(my_link)
assertTrue(my_link == grafeas_occurrence.to_link())
```

SantiagoTorres commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,277 @@
+  def __init__(self, server_url):
+    self.server_url = server_url
+
+  def dispatch(self, in_toto_link, note_name, resource_uri):
```
This is following the signature from the jenkins transports right? Is this a conscious decision? I wonder if instead we'd like to have something that follows the CRUD api nomenclature from the Grafeas server:

```
transport = GrafeasInTotoTransport("https://nyu.wtf")
transport.create(my_link)
my_link = transport.read(name)
transport.update(name, my_link)
transport.delete(name)
```

adityasaky replied:
Do you have any thoughts on opting to make the transport a stand-alone PR?

ETA: I think the suggestion in the review is sensible, I think I originally wrote the signature to maintain consistency for the few transports we have across our projects.

SantiagoTorres commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,277 @@
+    if "sig" not in signature:
+      signatures.append({"keyid": signature["keyid"], "sig": signature["signature"]})
+    else:
+      signatures.append(signature)
```
would it make sense to `validate()` the link at some point here?

adityasaky replied:
Links are now validated when `from_link` is first called. Great suggestion!

SantiagoTorres reviewed (commented):
Cool stuff!

I have mostly cosmetic comments, I hope they help!

kristelfung committed:
Update in-toto-verify unit test

Signed-off-by: Kristel Fung <kristelfung@gmail.com>

adityasaky committed:
Move to from_link

Signed-off-by: Aditya Sirish <aditya@saky.in>

kristelfung committed:
- Add/update test metadata
- Move create functions into occurrence class
- Single layout in-toto-verify - setup a full supply chain
- Check for Grafeas specific "custom_values"

Signed-off-by: Kristel Fung <kristelfung@gmail.com>
Signed-off-by: Aditya Sirish <aditya@saky.in>

kristelfung committed:
Move occurrences and links to folders, remove clone_note.json

Signed-off-by: Kristel Fung <kristelfung@gmail.com>

adityasaky committed:
- Code style clean up
- Remove Grafeas Transport - out of scope for this PR

Signed-off-by: Aditya Sirish <aditya@saky.in>

kristelfung committed:
Clean up test code

Signed-off-by: Kristel Fung <kristelfung@gmail.com>
Signed-off-by: Aditya Sirish <aditya@saky.in>

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+from in_toto.models.link import Link
+from in_toto.models.metadata import Metablock
+import requests
+import json
```
Minor nit: Please follow [styleguide on "Imports formatting"](https://google.github.io/styleguide/pyguide.html#313-imports-formatting). In test module too.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+  products = None
+  command = None
+  byproducts = None
+  environment = None
```
Note that these are class variables, which is probably not what you want, since you override them with instance variables below in `__init__`.  I suggest to take a couple of minutes to read up on class vs. instance variables in Python.

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,161 @@
+    grafeas_occurrence = \
+        GrafeasInTotoOccurrence.load("occurrences/clone_occurrence.json")
+    in_toto_link = grafeas_occurrence.to_link("clone-step-name")
+    assert isinstance(in_toto_link, Metablock)
```
Please use unittest's `self.assert*` instead of `assert` here and below.

lukpueh replied:
Also, the test would be a lot stronger if you asserted that `to_link` returns not just any `Metablock` but exactly the one you want, which you do have available in your test files (i.e. `links/clone_link.json`).

Same goes for the `from_link` test below.

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,161 @@
+    try:
+      os.remove("in-toto-verify-links/package.2f89b927.link")
+    except OSError:
+      pass
```
Minor nit: Looks like a job for a `for` loop.

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,161 @@
+    grafeas_occurrence_package = \
+        GrafeasInTotoOccurrence.load("occurrences/package_occurrence.json")
+    in_toto_link_package = grafeas_occurrence_package.to_link("package")
+    in_toto_link_package.dump("in-toto-verify-links/package.2f89b927.link")
```
Minor nit: This too looks like a job for a for loop. :)

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,161 @@
+
+    in_toto_verify(self.metadata,
+                   self.layout_key_dict,
+                   link_dir_path="in-toto-verify-links")
```
Minor nit: Tests are usually stronger if they assert that an operations yields the expected result. I assume that you implicitly assert here that `in_toto_verify` does not error out. I would at least add a comment about that. Or you could also take a look at the return value of `in_toto_verify` and make an assertion about it, to make your test stronger.

adityasaky replied:
`in_toto_verify` returns a summary link (for sublayout support). Rather, I think wrapping `in_toto_verify` in `assertTrue` will suffice? `assertTrue(expr)` checks that `bool(expr) == True`. Or would you prefer if I didn't do that?

lukpueh replied:
I guess a comment is just as good.

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,161 @@
+
+
+class TestCreateOccurranceFromInTotoLink(unittest.TestCase):
+  """Tests conversion of a in-toto link to a Grafeas occurrence."""
```
```suggestion
  """Tests conversion of an in-toto link to a Grafeas occurrence."""
```

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+        "products": self.products,
+        "environment": self.environment,
+        "byproducts": self.byproducts
+    })
```
Why not re-use `to_dict`?

lukpueh replied:
And for consistency with `GrafeasInTotoOccurrence`, I'd probably call this `to_json` too.

adityasaky replied:
`GrafeasInTotoOccurrence.to_json` returns JSON while this creates a dictionary... I'm worried about confusion in what's returned, there.

adityasaky replied:
I've added to_dict to GrafeasInTotoOccurrence to avoid confusion and this is now used within to_json. This is likely a good thing regardless, to have access to that representation.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+  intoto = {
+    "signatures":  [],
+    "signed": None
+  }
```
See comment about class vs. instance variables above (!!).

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+            "signed": self.intoto["signed"].to_dict()
+          }
+        }
+      )
```
You can deduplicate code here, by `json.dumps`ing only once and then writing the contents to file depending on the presence of a `file_path`. Also, I'd probably return the string in any case, at least if you keep the docstring as it is now.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+    if materials is None:
+      self.materials = []
+    elif isinstance(materials, list):
+      self.materials = materials
```
What happens if the passed `materials` is neither `None`, nor a `list` instance? Given that `materials` is a class variable, this can have particularly surprising results. 

I actually recommend you to try creating two `GrafeasLink` objects in a row, passing a non-None/non-list materials argument, and then inspect both of them.

This comment also applies to the rest of the constructor.

adityasaky replied:
I cleaned up the constructor a bit and wrote another "helper" to step in for how we were using the constructor previously. Now the constructor only creates a new GrafeasLink instance. Is this better?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+        grafeas_link.byproducts["custom_values"][key] = value
+
+    for key, value in link.environment.items():
+      grafeas_link.environment["custom_values"][key] = value
```
Can you comment the two for loops a bit? What are `custom_values`? Why do you case-handle `"return-value"`? etc.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+    kind: Set to the string "INTOTO" to indicate the type of occurrence.
+    resource: A dict of resource uri.
+    intoto: A dictionary with key "signatures", a list of signatures consisting
+        of key id and signature hash, and key "signed", the GrafeasLink class.
```
Any reason why this is a dict and not two fields?

lukpueh replied:
ping

adityasaky replied:
Are you suggesting we have an intoto_signatures and an intoto_signed field here? I used a dictionary to maintain the levels as they occur in the occurrences. I suppose the other option is to set up a class to have intoto.signed and intoto.signatures?

lukpueh replied:
I think the dict is fine.  The "right" value for `intoto` would probably be a `GrafeasInTotoMetadata` instance, with `signatures: list` and `signed: GrafeasLink` fields. But that seems overkill. What actually was the motivation for creating a `GrafeasLink` class? Wouldn't an `intoto["signed"]` dictionary work just as well for the purpose of de/serialization?

adityasaky replied:
The idea was to separate out the Grafeas wrapper from the in-toto fields initially. Do you think it should be merged in? I'm not a huge fan of the current mix of dictionary and object notation.

lukpueh replied:
What's the expected lifetime of these objects? If they are only instantiated to convert between links and occurrences, i.e. `occurrence-name.json <-> GrafeasInTotoOccurrence[GrafeasLink] <-> step-name.link`, then I don't see a huge benefit of a complex class model. OTOH, if users are supposed to do some work with those objects (read/write attributes, etc.) then the classes can indeed be helpful.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+  kind = "INTOTO"
+  resource = {
+    "uri": None
+  }
```
Why the default, if you override it in the constructor?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+              "signatures": self.intoto["signatures"],
+              "signed": self.intoto["signed"].to_dict()
+            }
+          }, fp, indent=4)
```
Why `indent=4` here but not below?

adityasaky replied:
Right now, `indent=4` is only set when dumping out to a file. Should I get rid of it altogether?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+    elif isinstance(environment, dict):
+      self.environment = environment
+
+  @staticmethod
```
I think it's common practice to use class methods for factories. (I recommend to read up on staticmethod vs. classmethod)

lukpueh replied:
ping. disagree?

adityasaky replied:
Just updated this, sorry. Should `load` also be a classmethod? I've retained it as a staticmethod for now, it's also a staticmethod for Metablock.load().

lukpueh replied:
> Should load also be a classmethod?

It's not a big deal but it makes the factory slightly more self-contained, e.g.:

```python
class Foo:
  @staticmethod
  def create():
    return Foo() # <- needs knowledge about Foo from outer scope

 # vs.
class Foo:
  @classmethod
  def create(cls):
    return cls() # <- no need to leave local scope + survives rename
```

> it's also a staticmethod for Metablock.load().

Oh, I probably didn't know that back when I wrote `Metablock.load` :)


This [realpython tutorial](https://realpython.com/instance-class-and-static-methods-demystified/) nicely explains the differences.

adityasaky replied:
Thanks for the link! I've updated `load` as well.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+      for key, value in \
+          self.intoto["signed"].byproducts["custom_values"].items():
+        if key == "return-value":
+          byproducts[key] = int(value)  # cast return-value back to int
```
Important rule of commenting: Don't state the obvious! :) It should be clear from the code that we cast to int, but the reader can only guess why.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+    command = self.intoto["signed"].command
+
+    for key, value in self.intoto["signed"].byproducts.items():
+      if key == "custom_values":
```
Maybe add a comment that we will handle this case below (and also why we handle it below and not here). Also, do we expect any other values?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+        signatures.append({"keyid": signature["keyid"],
+                           "sig": signature["signature"]})
+      else:
+        signatures.append(signature)
```
Do we expect a list of signatures with mixed ("sig" and "signature") keys?

adityasaky replied:
I'm not sure we'd have mixed instances, but it could be either technically, given my understanding of this: https://github.com/grafeas/grafeas/blob/master/proto/v1beta1/intoto.proto#L101

lukpueh replied:
I don't know protobuf but if this line implies that the wire format can have either "sig" or "signature" fields, then we'd also need to account for ["keyid" and "key_id"](https://github.com/grafeas/grafeas/blob/eab9b085da7a038510999e1e1b3bf630926254e5/proto/v1beta1/intoto.proto#L100), or ["signed" and "link"](https://github.com/grafeas/grafeas/blob/eab9b085da7a038510999e1e1b3bf630926254e5/proto/v1beta1/intoto.proto#L95). Or am I missing something?

adityasaky replied:
So with some testing, I'm beginning to think that it doesn't matter - it seems to default to the field name in the json_name field, even if my input has the actual field. Therefore, I've simplified things a bit across this layer.

adityasaky replied:
I've also updated the test occurrence metadata - I regenerated them using the link files and I was able to briefly test them using a local Grafeas instance. Perhaps @kristelfung could check my work and test the new metadata against a Grafeas server as well? :)

lukpueh commented on tests/occurrences/update_occurrence.json:
```diff
@@ -0,0 +1,40 @@
+            }
+        }
+    }
+}
```
Why is the indentation in this occurrence different from the occurrences?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+      self.intoto["signatures"] = []
+      for signature in in_toto_link.signatures:
+        self.intoto["signatures"].append({"keyid": signature["keyid"],
+                                          "signature": signature["sig"]})
```
Minor design nit: For the sake of consistency, I'd probably do the link-to-occurrence conversion in the `from_link` factory and not in the constructor (similar to how you do it in `GrafeasLink`), because this constructor is also used by `load`, which expects an occurrence file that does not require any conversion. Do you see what I mean?

adityasaky replied:
Good point! The constructor now merely returns a new instance of `GrafeasInTotoOccurrence`.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+          byproducts[key] = value
+
+    for key, value in self.intoto["signed"].environment.items():
+      if key == "custom_values":
```
Maybe add a comment that we will handle this case below (and also why we handle it below and not here). Also, do we expect any other values?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,266 @@
+      else:
+        signatures.append(signature)
+
+    return Metablock(signed=in_toto_link, signatures=signatures)
```
Is there a reason why all of the `to_link` work happens in a `GrafeasInTotoOccurrence` method, but most of the `from_link` work happens in a `GrafeasLink` method?

adityasaky replied:
I think this is better handled now - `GrafeasInTotoOccurrence.to_link` calls `GrafeasLink.to_link` for the fields that are relevant to the latter while handling the fields relevant to the former.

Similarly, `GrafeasInTotoOccurrence.from_link` relies on `GrafeasLink.from_link`. I think this is neater now?

lukpueh reviewed (commented):
Thanks for working on this, @kristelfung and @adityasaky! 

I'd love to see a bit more code documentation. In particular, I would like to know at one glance what properties need to change in a link-occurrence conversion. This code does a lot of copying (as is) from one data structure to another data structure, so it's a bit hard to see were the conversion happens (at least for me).

adityasaky committed:
Address review comments - more code style fixes

Signed-off-by: Aditya Sirish <aditya@saky.in>

kristelfung committed:
- Fix tests:
  - Fix indentation of metadata and regenerate where applicable
  - Use unittest's assert functions
  - Iterative loading of test metadata where appropriate
- Improve code structure (Lukas's review):
  - Use classmethods
  - Handle all conversions in GrafeasInTotoOccurrence

Signed-off-by: Kristel Fung <kristelfung@gmail.com>
Signed-off-by: Aditya Sirish <aditya@saky.in>

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,147 @@
+        GrafeasInTotoOccurrence.from_link(in_toto_link, "clone",
+                                          "clone-resource-uri")
+
+    assert isinstance(grafeas_occurrence, GrafeasInTotoOccurrence)
```
[ping](https://github.com/in-toto/totoify-grafeas/pull/3#discussion_r559479954)

lukpueh commented on tests/test_totoifylib.py:
```diff
@@ -0,0 +1,147 @@
+    """
+    for f in os.listdir("occurrences"):
+      occurrence = GrafeasInTotoOccurrence.load(os.path.join("occurrences", f))
+      # this is hacky, yes, but since we control the test metadata...
```
I think it's quite okay. 👍

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    for key, value in self.byproducts.items():
+      if key == "custom_values":
+        continue
+      byproducts[key] = value
```
Probably a matter of taste, but to me this seems easier to read:
```suggestion
      if key != "custom_values":
        byproducts[key] = value
```
Same comment applies below in `for key, value in self.environment.items()`-loop.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+      byproducts[key] = value
+
+    if "custom_values" in self.byproducts:
+      for key, value in \
```
I see no need to break the line here.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+      environment[key] = value
+
+    if "custom_values" in self.environment:
+      for key, value in \
```
I see no need to break the line here.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    if environment is None:
+      self.environment = {"custom_values": {}}
+    else:
+      self.environment = environment
```
I think a blank line between the `if/else` blocks facilitates readability.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    self.resource = {
+        "uri": resource_uri
+    }
+    self.kind = "INTOTO"
```
I think a blank line between the `if/else` blocks facilitates readability.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    products = dictionary['products']
+    command = dictionary['command']
+    byproducts = dictionary['byproducts']
+    environment = dictionary['environment']
```
There's a bit of a mix of single and double quotes in this document. Please consolidate! (double are preferred)

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    if command is None:
+      self.command = ""
+    else:
+      self.command = command
```
You could spare this `if/else` block and instead make "" the default kwarg in the function signature.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    if intoto is None:
+      self.intoto = {
+          "signatures": [],
+          "signed": None
```
The default `None` will be a problem if you call `to_dict` or `to_json`, which expects `self.intoto.signed` to have a `to_dict` method. Is there a use case for this class that requires default arguments?

adityasaky replied:
I'm using `GrafeasLink()` now. As for default arguments, I think not. The intended workflow for use is `.load`/`.to_link`/`.from_link` etc. Should this move to just compulsory arguments?

lukpueh replied:
If you only intend to use .load, .from_link and those methods only pass non-None values for all constructor parameters then I think you can make them compulsory. This would also spare you all of the `if <passed_arg> is None: ... / else: ...` boilerplate  code. :)

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    else:
+      self.note_name = note_name
+    if resource_uri is None:
+      resource_uri = ""
```
You could spare these `if/else` blocks and instead make "" the default kwarg in the function signature.

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    with open(path, "r") as occ_f:
+      occ_json = json.load(occ_f)
+
+    intoto = {"signatures": [], "signed": None}
```
Any reason to initialize "signatures" and "signed", if they get overwritten right away?

lukpueh commented on totoify_grafeas/totoifylib.py:
```diff
@@ -0,0 +1,269 @@
+    intoto = {"signatures": [], "signed": None}
+    intoto['signed'] = GrafeasLink.from_dict(occ_json['intoto']['signed'])
+    intoto['signatures'] = occ_json['intoto']['signatures']
+    return cls(intoto=intoto, resource_uri=occ_json['resource']['uri'], note_name=occ_json['noteName'])
```
line length

lukpueh reviewed (changes requested):
Thanks for addressing most of my comments, @adityasaky I think this is pretty much good to go. I only have a couple of more minor nits.

adityasaky commented:
@lukpueh thanks a ton for shepherding this PR! I've made some changes to address the comments re GrafeasLink and default values in the constructor. Let me know if this looks good to you.

@kristelfung we should probably squash commits on this branch, would you prefer if I did it?

lukpueh reviewed (approved):
I think you changes simplified the code quite a bit Thanks, @adityasaky, LGTM now!

kristelfung commented:
> @lukpueh thanks a ton for shepherding this PR! I've made some changes to address the comments re GrafeasLink and default values in the constructor. Let me know if this looks good to you.
> 
> 
> 
> @kristelfung we should probably squash commits on this branch, would you prefer if I did it?

Yes please! Sorry, I've been swamped recently!

adityasaky commented:
I've squashed it down to 15 commits, with authorship reasonably well defined :)

@lukpueh @SantiagoTorres I think this can be merged now.
